src/GafferUI/StandardNodule.cpp

```diff
@@ -465,11 +467,21 @@ void StandardNodule::setCompatibleLabelsVisible( const DragDropEvent &event, boo
 		return;
 	}
 
+	IECore::ConstBoolDataPtr inputVisibleData = Gaffer::Metadata::value<IECore::BoolData>( nodeGadget->node(), g_inputNoduleLabelsVisibleKey );
+	IECore::ConstBoolDataPtr outputVisibleData = Gaffer::Metadata::value<IECore::BoolData>( nodeGadget->node(), g_outputNoduleLabelsVisibleKey );
+
 	for( StandardNodule::RecursiveIterator it( nodeGadget ); !it.done(); ++it )
 	{
 		if( creator->canCreateConnection( it->get()->plug() ) )
 		{
-			(*it)->setLabelVisible( visible );
+			if( it->get()->plug()->direction() == Plug::Direction::In )
+			{
+				(*it)->setLabelVisible( !inputVisibleData ? visible : inputVisibleData->readable() );
+			}
+			if( it->get()->plug()->direction() == Plug::Direction::Out )
+			{
+				(*it)->setLabelVisible( !outputVisibleData ? visible : outputVisibleData->readable() );
+			}
```

Comment:
As per my original comments, I think we can use the original code here too, for the same reasons as before.

Reply:
Without this, when you `dragLeave()` or `drop()` on a nodule, all the node's nodules visibility will be set to `false` because we're unconditionally doing that in those events.

Maybe instead of this, we create a private `staticVisibilityState()` (or somesuch name) that we use instead of `false` for the `dragLeave()` and `drop()` events?

Reply:
Ah, OK. I had completely missed that `setCompatibleLabelsVisible()` was called for leave and drop as well as enter. Thanks for pointing that out.

> Maybe instead of this, we create a private staticVisibilityState() (or somesuch name) that we use instead of false for the dragLeave() and drop() events?

Yeah, or perhaps we just call `updatePersistentNoduleLabels()` on leave and drop? So the pattern is that helpful things set the visibility however they want, and then we always use `updatePersistentNoduleLabels()` to clobber everything back to the default state when we're done?

The problem with that of course is that `updatePersistentNoduleLabels()` is private to StandardNodeGadget at present. Which highlights how unsatisfactory the original code was in the first place, with the responsibility for visibility management spread out all over.

Maybe we just rename the private method to `StandardNodeGadget::applyNoduleLabelVisibilityMetadata()` and make it public?

Reply:
I went with the second solution in https://github.com/GafferHQ/gaffer/pull/6135/commits/a9f2a2fefeb86695074e5153a76ea1caa87fe6f4. And noting here our offline conversation where we decided to deprecate the current `updatePersistentNoduleLabels()` in favor of one that does not have the `visible` argument.

https://github.com/GafferHQ/gaffer/pull/6135/commits/a9f2a2fefeb86695074e5153a76ea1caa87fe6f4 maintains backward compatibility in case there are uses of `StandardNodule` we don't know about. Some of that can (must) be removed when we remove the deprecated `updatePersistentNoduleLabels()`.